### ISSUE TITLE
yolo26x-seg training

### ISSUE THREAD

**Issue by Wx-create** [ultralytics/ultralytics#23622]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### Ultralytics YOLO Component

Train

### Bug

在yolo26-seg实例分割模型中计算semseg_loss时，sem_masks的尺寸和pred_semseg的尺寸不一致。例如：若输入尺寸为640，sem_masks的尺寸为160*160，pred_semseg的尺寸为80*80

### Environment

Ultralytics 8.4.7 🚀 Python-3.8.5 torch-2.0.0+cu118 CUDA:0 (NVIDIA GeForce RTX 4090 D, 24210MiB)
Setup complete ✅ (384 CPUs, 1007.5 GB RAM, 3082.6/7392.3 GB disk)

OS                     Linux-5.15.0-164-generic-x86_64-with-glibc2.10
Environment            Docker
Python                 3.8.5
Install                pip
Path                   /home/zhengzhiwei/ultralytics-main/ultralytics
RAM                    1007.50 GB
Disk                   3082.6/7392.3 GB
CPU                    AMD EPYC 9654 96-Core Processor
CPU count              384
GPU                    NVIDIA GeForce RTX 4090 D, 24210MiB
GPU count              8
CUDA                   11.8

numpy                  ✅ 1.23.4>=1.23.0
matplotlib             ✅ 3.5.2>=3.3.0
opencv-python          ✅ 4.13.0.92>=4.6.0
pillow                 ✅ 9.4.0>=7.1.2
pyyaml                 ✅ 6.0>=5.3.1
requests               ✅ 2.24.0>=2.23.0
scipy                  ✅ 1.9.1>=1.4.1
torch                  ✅ 2.0.0+cu118<2.10,>=1.8.0
torchvision            ✅ 0.15.1+cu118>=0.9.0
psutil                 ✅ 6.1.0>=5.8.0
polars                 ✅ 1.8.2>=0.20.0
ultralytics-thop       ✅ 2.0.18>=2.0.18
torch                  ✅ 2.0.0+cu118!=2.4.0,<2.10,>=1.8.0; sys_platform == "win32"

### Minimal Reproducible Example

# 创建 ArgumentParser 对象
parser = argparse.ArgumentParser(description="模型训练参数设置")

# 添加参数
parser.add_argument("--model_config", type=str, help="模型参数文件路径")
parser.add_argument("--data_config", type=str, help="数据所在路径")
parser.add_argument("--imgsz", type=int, help="图像输入模型尺寸")
parser.add_argument("--epochs", type=int, help="模型训练的epoch")
parser.add_argument("--batch_size", type=int, help="模型训练的batch_size")
parser.add_argument("--pretrained_path", type=str, help="预训练模型所在路径")
parser.add_argument("--gpu", type=str, help="训练使用的gpu卡号")
parser.add_argument("--save_dir", type=str, help="训练后的模型保存路径")
parser.add_argument("--proj_name", type=str, help="项目训练的模型名称")
    

# 解析参数
args = parser.parse_args()

model = YOLO(model=args.model_config)
model.load(args.pretrained_path)
model.train(data=args.data_config,
                imgsz=args.imgsz,
                epochs=args.epochs,
                save_period=1,
                batch=args.batch_size,
                workers=0,
                device=args.gpu,
                optimizer="AdamW",
                lr0=0.001,
                cos_lr=True,
                patience=10000,
                close_mosaic=100,
                resume=False,
                project=args.save_dir,
                name=args.proj_name,
                single_cls=False,
                cache=False,
                plots=True,
                amp=False,
                translate=0.,
                scale=0.,
                mosaic=0.,
                erasing=0.
                )

#loss计算（v8SegmentationLoss）
#计算semseg，输入的真值和预测值尺寸不一致
loss[4] = self.bcedice_loss(pred_semseg, sem_masks)

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @Wx-create, thank you for your interest in Ultralytics 🚀! We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

This is an automated response 🤖—an Ultralytics engineer will also assist soon.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. You’ve already shared a great start—could you please extend it to include the *smallest runnable snippet* that reproduces the `sem_masks` vs `pred_semseg` shape mismatch (including the exact `model_config`, `data_config` structure, and a single sample/mini dataset setup)? Also include the full traceback/log lines around the loss computation and the printed tensor shapes right before `self.bcedice_loss(pred_semseg, sem_masks)` 📌

If this is a custom training ❓ Question, please provide as much information as possible, including dataset image examples and training logs, and verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo26"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by Y-T-G** [comment]
@lmycross

**Comment by lmycross** [comment]
The sem_masks are generated based on the instance masks, so they share the same spatial size, while the size of pred_semseg is fixed, as it corresponds to the input resolution downsampled by a factor of 8. Therefore, their spatial sizes are different. During loss computation, the sem_masks are resized to match the size of pred_semseg.
https://github.com/ultralytics/ultralytics/blob/004deb98f0462f06c925680e0538326c82722917/ultralytics/utils/loss.py#L306-L311
Are you encountering any error? If so, could you share the full error message?

**Comment by Wx-create** [comment]
在BCEDiceLoss中确实有判断预测值和真值不一致的情况并进行处理。还有一点，我想请问，为什么yolo26实例分割模型在一个数据量较大的数据集上训练（42w），在测试集上的实例掩膜预测并不好，掩膜并不贴近目标边缘，而是占满了目标的box框。但是在一个小的子数据集上训练（1w5），在测试机上的实例掩膜预测会正常。训练参数实验配置都是一样。迭代epoch为60，end2end设置为True，会是end2end训练模式不易拟合的原因吗？

**Comment by lmycross** [comment]
Here are a few suggestions to investigate:
1. Check data consistency between the small and large datasets and make sure annotation format, mask quality, and preprocessing steps are identical. If masks in large dataset are coarse or close to rectangular, the model may learn to fill the whole box.
2. Check training curves. If both training and validation mask loss are still clearly decreasing at epoch 60, the model is likely undertrained. In this case, you should increase epochs. If training mask loss keeps decreasing but validation mask loss plateaus or increases, that suggests overfitting or label noise.

**Comment by Wx-create** [comment]
1. 大数据集送入训练过程中，保存参与训练的样本，mask都是贴近实例的，并不是充满box框，实例都是包裹，会有类似矩形形状，也有其他形状。
2. 当mosaic增强关闭后，在最后10个epoch训练时loss会再降低，seg_loss在最后10个epoch都维持在0.32左右。
我尝试增加到100epoch试试。

**Comment by glenn-jocher** [comment]
如果你在关闭 `mosaic` 的最后 10 个 epoch 还能继续下降但 `seg_loss` 长时间卡在 ~0.32，建议你先做两个快速 A/B：把 `imgsz` 提高一些（例如 960/1280）看边缘是否变得更贴合；同时把训练时的 `box` 增大一点以提高分割损失权重（当前实现里 `seg`/`semseg` 都是乘 `hyp.box` 的，见 [v8SegmentationLoss.loss 实现](https://docs.ultralytics.com/reference/utils/loss/)），另外再用非 `end2end` 的同尺寸分割配置/权重跑一版对照排除 `end2end` 影响；如果 100 epoch 后仍然“填满 box”，麻烦再贴一下 `runs/segment/.../results.csv` 和几张验证集 GT/Pred 叠图，我们好判断是欠拟合还是标签噪声/尺度问题（训练入口逻辑可参考 [SegmentationTrainer](https://docs.ultralytics.com/reference/models/yolo/segment/train/)）。
